### PR TITLE
merge upstream

### PR DESCRIPTION
**Pull request by antoinepairet**

### CODE REVIEW

**Comment by cebamps** [comment]
@antoinepairet Gros changements dans la crypto, au-dela de ceux que j'ai contribués upstream. Il faudra tester ça proprement. Veux-tu que je regarde en détail ce qui a changé?

Aussi, le changement des `utiDefs` du document-x-api pourrait avoir une incidence sur la détection du type de fichier des pièces jointes de la eHealthBox pour ce qui est des fichiers plaintext (donc H1 et Medidoc compris). Il faudra tester proprement les messages pre- et post-merge pour voir si les types sont bien détectés. A priori ça va casser un fix dans notre routine `extractDocuments`.

**Comment by cebamps** [comment]
@antoinepairet The latest changes seem to add more crypto robustness. Not reviewed yet.

@fthuin Can you check recent changes in invoice batch creation in commit 632bb403a2fcc236f05872a7f9691d30135bab70 ?

**Comment by fthuin** [comment]
I don't think we have a concept of financialInstitutionInformation in Medispring so this would probably break eFact.

File: icc-api/api/XHR.ts [hunk 1]
```diff
@@ -75,11 +75,11 @@ export namespace XHR {
         throw new XHRError(response.statusText, response.status, response.status, response.headers)
       }
       const ct = response.headers.get("content-type") || "text/plain"
-      return (ct.startsWith("application/octet-stream")
-        ? response.arrayBuffer()
-        : ct.startsWith("application/json")
-          ? response.json()
-          : response.text()
+      return (ct.startsWith("application/json")
```
Comment (phil1618) [inline icc-api/api/XHR.ts]: Possible risk: wring mime type expected. Example: MyCarenetConversation list.

File: icc-x-api/utils/fhc-invoice-sender-util.ts [hunk 2]
```diff
@@ -2,19 +2,30 @@ import { HealthcarePartyDto } from "../../icc-api/model/models"
...
+  const iban = hcp.bankAccount || (fi && fi.bankAccount)
+
+  if (!bic || !iban) {
+    throw new Error("Missing bic or iban for HCP")
+  }
+
   const invoiceSender: InvoiceSender = {
     bce: Number(hcp.cbe),
-    bic: hcp.bic,
+    bic: bic!.replace(/ /g, ""),
```
Comment (phil1618) [inline icc-x-api/utils/fhc-invoice-sender-util.ts]: Why not adding [^A-Z0-9] ? (keep only numerical value)

File: icc-x-api/utils/fhc-invoice-sender-util.ts [hunk 3]
```diff
@@ -2,19 +2,30 @@ import { HealthcarePartyDto } from "../../icc-api/model/models"
...
     conventionCode: hcp.convention,
     firstName: hcp.firstName,
-    iban: hcp.bankAccount,
+    iban: iban!.replace(/ /g, ""),
```
Comment (phil1618) [inline icc-x-api/utils/fhc-invoice-sender-util.ts]: idem than previous comment

File: icc-api/api/iccCodeApi.ts [hunk 4]
```diff
@@ -55,6 +55,37 @@ export class iccCodeApi {
       .then(doc => new models.CodeDto(doc.body as JSON))
       .catch(err => this.handleError(err))
   }
+  filterBy(
```
Comment (phil1618) [inline icc-api/api/iccCodeApi.ts]: Should be used in MS too in combination with frequency count.

File: icc-api/api/iccUserApi.ts [hunk 5]
```diff
@@ -127,6 +127,18 @@ export class iccUserApi {
       .then(doc => new models.UserDto(doc.body as JSON))
       .catch(err => this.handleError(err))
   }
+  getMatchingUsers(): Promise<Array<models.UserDto> | any> {
```
Comment (phil1618) [inline icc-api/api/iccUserApi.ts]: Should be used to detect login/password duplicated user. If the number of matching user is bigger than 1 the user should be blocked and contact the helpdesk asap

File: icc-api/model/CalendarItem.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
+/**
```
Comment (phil1618) [inline icc-api/model/CalendarItem.ts]: This is a regression as only the CalendarItemDto should be exposed in the icc-api. The file should be deleted (new genloc required)

File: icc-api/model/Valorisation.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (phil1618) [inline icc-api/model/Valorisation.ts]: Check if is not a overshoot regarding the already exposed  ValorisationDto.

File: icc-x-api/icc-bekmehr-x-api.ts [hunk 8]
```diff
@@ -23,14 +23,7 @@ export class IccBekmehrXApi extends iccBekmehrApi {
     const auth = this.headers.find(h => h.header === "Authorization")
     this.wssHost = new URL(this.host, window.location.href).href
       .replace(/^http/, "ws")
-      .replace(/:\/\//, "://" + (auth ? atob(auth.data.replace(/Basic /, "")) + "@" : ""))
```
Comment (phil1618) [inline icc-x-api/icc-bekmehr-x-api.ts]: Correction needed to comply with Chrome limitation. The identification is handled by the sessionId specified below

File: icc-x-api/icc-document-x-api.ts [hunk 9]
```diff
@@ -122,32 +122,32 @@ export class IccDocumentXApi extends iccDocumentApi {
...
-    "public.mpeg4Audio": "audio/mp4",
+    "public.mpeg4-audio": "audio/mp4",
     "public.mpegts": "video/MP2T",
-    "public.objectiveCPlusPlusSource": "text/x-objcppsrc",
-    "public.objectiveCSource": "text/x-objcsrc",
-    "public.opentypeFont": "font/opentype",
-    "public.perlScript": "text/x-perl",
-    "public.phpScript": "text/php",
-    "public.plainText": "text/plain",
+    "public.objective-c-plus-plus-source": "text/x-objcppsrc",
```
Comment (phil1618) [inline icc-x-api/icc-document-x-api.ts]: Section to check on MS side

File: icc-x-api/icc-invoice-x-api.ts [hunk 10]
```diff
@@ -160,12 +160,20 @@ export class IccInvoiceXApi extends iccInvoiceApi {
     prefix: string,
     entityrefApi: iccEntityrefApi
   ): Promise<models.EntityReference> {
-    return entityrefApi.createEntityReference(
-      new models.EntityReference({
-        id: prefix + nextReference.toString().padStart(6, "0"),
-        docId
+    return entityrefApi
```
Comment (phil1618) [inline icc-x-api/icc-invoice-x-api.ts]: Improve the entityReference uniqueness

File: icc-api/api/XHR.ts (hunk 1 above)
Comment (nicolas-b12) [reply icc-api/api/XHR.ts]: I don't see an issue here.

File: icc-api/api/iccIcureApi.ts [hunk 11]
```diff
@@ -154,6 +154,23 @@ export class iccIcureApi {
       .then(doc => true)
       .catch(err => this.handleError(err))
   }
+  resolveDocumentsConflicts(ids?: string): Promise<any | Boolean> {
```
Comment (cebamps) [inline icc-api/api/iccIcureApi.ts]: Should be used to resolve truncated attachments in messages.

File: icc-api/api/iccUserApi.ts (hunk 5 above)
Comment (cebamps) [reply icc-api/api/iccUserApi.ts]: This was removed upstream: https://github.com/taktik/icure-backend/commit/8908d37a220a627dd7db914782d2c8b300d29c6e

File: icc-x-api/icc-bekmehr-x-api.ts (hunk 8 above)
Comment (cebamps) [reply icc-x-api/icc-bekmehr-x-api.ts]: Not an issue in MS.

File: icc-x-api/icc-message-x-api.ts [hunk 12]
```diff
@@ -1043,8 +1070,7 @@ export class IccMessageXApi extends iccMessageApi {
     const year = moment().year()
 
     return getFederaton(invoices, this.insuranceApi).then(fed => {
-      const fedCode = fed.code === "306" ? "300" : fed.code
-      const prefix = `efact:${hcp.id}:${year}:${fedCode}:`
+      const prefix = `efact:${hcp.id}:${fed.code === "306" ? "300" : fed.code}:`
```
[the remark this answers is not in the document]
Comment (nicolas-b12) [reply icc-x-api/icc-message-x-api.ts]: The removal of the year is a regression I think, I don't see any reason to remove it

[the remark this answers is not in the document]
Comment (nicolas-b12) [reply icc-x-api/icc-message-x-api.ts]: The removal of the year is a regression I think, I don't see any reason to remove it

**Pull request closed without merging**